File: app/src/main/java/com/snaptiongame/app/data/models/Game.java
package com.snaptiongame.app.data.models;

import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * @author Tyler Wong
 */
public class Game {

    @SerializedName(ID)
    public int id;
    @SerializedName(USER_ID)
    public int userId;
    @SerializedName(START_DATE)
    public long startDate;
    @SerializedName(END_DATE)
    public long endDate;
    @SerializedName(IS_PUBLIC)
    public boolean isPublic;
    @SerializedName(IMG_TYPE)
    public String type;
    @SerializedName(CREATOR_NAME)
    public String creatorName;
    @SerializedName(IMAGE_URL)
    public String creatorImage;
    @SerializedName(CREATOR_ID)
    public int creatorId;
    @SerializedName(PICTURE)
    public String picture;
    @SerializedName(IMAGE_URL)
    public String imageUrl;
    @SerializedName(IMAGE_WIDTH)
    public int imageWidth;
    @SerializedName(IMAGE_HEIGHT)
    public int imageHeight;
    @SerializedName(TOP_CAPTION)
    public Caption topCaption;
    @SerializedName(USERS)
    public List<User> users;
    @SerializedName(TAGS)
    public List<Tag> tags;
    public List<String> sendTags;
    @SerializedName(FRIENDS)
    public List<Integer> friendIds;
    @SerializedName(BEEN_UPVOTED)
    public boolean beenUpvoted;
    @SerializedName(BEEN_FLAGGED)
    public boolean beenFlagged;
    @SerializedName(GAME_DURATION)
    public long gameDuration;
    @SerializedName(NUM_UPVOTES)
    public int numUpvotes;
    @SerializedName(IS_CLOSED)
    public boolean isClosed;

    public static final String ID = "id";
    public static final String USER_ID = "userId";
    public static final String START_DATE = "creationDate";
    public static final String END_DATE = "completedDate";
    public static final String IS_PUBLIC = "isPublic";
    public static final String PICKER = "picker";
    public static final String CREATOR_ID = "creator_id";
    public static final String CREATOR_NAME = "picker_name";
    public static final String IMG_TYPE = "imgType";
    public static final String PICTURE = "picture";
    public static final String IMAGE_URL = "url";
    public static final String IMAGE_WIDTH = "width";
    public static final String IMAGE_HEIGHT = "height";
    public static final String TOP_CAPTION = "topCaption";
    public static final String FITB_BEFORE = "fitbBefore";
    public static final String FITB_AFTER = "fitbAfter";
    public static final String TAGS = "tags";
    public static final String FRIENDS = "invitedUsers";
    public static final String USERS = "users";
    public static final String BEEN_UPVOTED = "beenUpvoted";
    public static final String BEEN_FLAGGED = "beenFlagged";
    public static final String GAME_DURATION = "gameDuration";
    public static final String NUM_UPVOTES = "numUpvotes";
    public static final String CREATOR_IMAGE = "creatorImage";
    public static final String IS_CLOSED = "isClosed";


    public Game() {

    }

    public Game(int userId, boolean isPublic, String picture, String type,
                List<String> tags, List<Integer> friendIds, long gameDuration) {
        this.userId = userId;
        this.isPublic = isPublic;
        this.picture = picture;
        this.type = type;
        this.sendTags = tags;
        this.friendIds = friendIds;
        this.gameDuration = gameDuration;
    }

    public Game(int id, long startDate, boolean isPublic, int creatorId, String picture, String type) {
        this.id = id;
        this.startDate = startDate;
        this.isPublic = isPublic;
        this.creatorId = creatorId;
        this.picture = picture;
        this.type = type;
    }

    @Override
    public boolean equals(Object obj) {
        return obj.getClass() == Game.class
                && this.id == ((Game) obj).id
                && this.beenUpvoted == ((Game) obj).beenUpvoted
                && this.beenFlagged == ((Game) obj).beenFlagged
                && ((this.topCaption == null && ((Game) obj).topCaption == null)
                || (this.topCaption != null && this.topCaption.equals(((Game) obj).topCaption)));
    }
}

File: app/src/main/java/com/snaptiongame/app/data/models/Game.kt
package com.snaptiongame.app.data.models

import com.google.gson.annotations.SerializedName

/**
 * @author Tyler Wong
 */
class Game {

    @SerializedName(ID)
    var id: Int = 0
    @SerializedName(USER_ID)
    var userId: Int = 0
    @SerializedName(START_DATE)
    var startDate: Long = 0
    @SerializedName(END_DATE)
    var endDate: Long = 0
    @SerializedName(IS_PUBLIC)
    var isPublic: Boolean = false
    @SerializedName(IMG_TYPE)
    var type: String? = null
    @SerializedName(CREATOR_NAME)
    var creatorName: String? = null
    @SerializedName(IMAGE_URL)
    var creatorImage: String? = null
    @SerializedName(CREATOR_ID)
    var creatorId: Int = 0
    @SerializedName(PICTURE)
    var picture: String? = null
    @SerializedName(IMAGE_URL)
    var imageUrl: String? = null
    @SerializedName(IMAGE_WIDTH)
    var imageWidth: Int = 0
    @SerializedName(IMAGE_HEIGHT)
    var imageHeight: Int = 0
    @SerializedName(TOP_CAPTION)
    var topCaption: Caption? = null
    @SerializedName(USERS)
    var users: List<User>? = null
    @SerializedName(TAGS)
    var tags: List<Tag>? = null
    var sendTags: List<String>? = null
    @SerializedName(FRIENDS)
    var friendIds: List<Int>? = null
    @SerializedName(BEEN_UPVOTED)
    var beenUpvoted: Boolean = false
    @SerializedName(BEEN_FLAGGED)
    var beenFlagged: Boolean = false
    @SerializedName(GAME_DURATION)
    var gameDuration: Long = 0
    @SerializedName(NUM_UPVOTES)
    var numUpvotes: Int = 0
    @SerializedName(IS_CLOSED)
    var isClosed: Boolean = false
    @SerializedName(GAME_ID)
    var gameId: Int = 0

    var isFromAnotherGame: Boolean = false

    constructor() {

    }

    // Constructor for creating Game with new picture
    constructor(isPublic: Boolean, picture: String, type: String, tags: List<String>,
                friendIds: List<Int>, gameDuration: Long) {
        this.isPublic = isPublic
        this.picture = picture
        this.type = type
        this.sendTags = tags
        this.friendIds = friendIds
        this.gameDuration = gameDuration
        this.isFromAnotherGame = false
    }

    // Constructor for creating Game from gameId
    constructor(gameId: Int, isPublic: Boolean, tags: List<String>, friendIds: List<Int>,
                gameDuration: Long) {
        this.gameId = gameId
        this.isPublic = isPublic
        this.sendTags = tags
        this.friendIds = friendIds
        this.gameDuration = gameDuration
        this.isFromAnotherGame = true
    }

    // Constructor for Game from server
    constructor(id: Int, startDate: Long, isPublic: Boolean, creatorId: Int, picture: String, type: String) {
        this.id = id
        this.startDate = startDate
        this.isPublic = isPublic
        this.creatorId = creatorId
        this.picture = picture
        this.type = type
    }

    override fun equals(other: Any?): Boolean {
        return other is Game
                && this.id == other.id
                && this.beenUpvoted == other.beenUpvoted
                && this.beenFlagged == other.beenFlagged
                && (this.topCaption == null && other.topCaption == null || this.topCaption != null
                && this.topCaption == other.topCaption)
    }

    companion object {
        const val ID = "id"
        const val USER_ID = "userId"
        const val START_DATE = "creationDate"
        const val END_DATE = "completedDate"
        const val IS_PUBLIC = "isPublic"
        const val PICKER = "picker"
        const val CREATOR_ID = "creator_id"
        const val CREATOR_NAME = "picker_name"
        const val IMG_TYPE = "imgType"
        const val PICTURE = "picture"
        const val IMAGE_URL = "url"
        const val IMAGE_WIDTH = "width"
        const val IMAGE_HEIGHT = "height"
        const val TOP_CAPTION = "topCaption"
        const val FITB_BEFORE = "fitbBefore"
        const val FITB_AFTER = "fitbAfter"
        const val TAGS = "tags"
        const val FRIENDS = "invitedUsers"
        const val USERS = "users"
        const val BEEN_UPVOTED = "beenUpvoted"
        const val BEEN_FLAGGED = "beenFlagged"
        const val GAME_DURATION = "gameDuration"
        const val NUM_UPVOTES = "numUpvotes"
        const val CREATOR_IMAGE = "creatorImage"
        const val IS_CLOSED = "isClosed"
        const val GAME_ID = "gameId"
    }
}
